### PR TITLE
add code to trigger after_send_error event

### PR DESCRIPTION
**Pull request by junchaowu**
Somehow the after_send_error event is missing from current master. Add it.
Add tracing to the error too.

### CODE REVIEW

File: tchannel/errors.py [hunk 1]
```diff
@@ -77,7 +79,7 @@ def __init__(
         tracing=None,
     ):
         super(TChannelError, self).__init__(description)
-        self.tracing = tracing
+        self.tracing = tracing or Trace()
```
Comment (blampe) [inline tchannel/errors.py]: Huh? Why?

File: tchannel/tornado/connection.py [hunk 2]
```diff
@@ -471,25 +470,21 @@ def serve(self, handler):
                 # TODO Send error frame back
                 logging.exception("Failed to process %s", repr(message))
 
-    def send_error(self, code, description, message_id):
+    def send_error(self, error):
```
Comment (blampe) [inline tchannel/tornado/connection.py]: This is a breaking change, shouldn't land in a patch.

File: tchannel/errors.py (hunk 1 above)
Comment (junchaowu) [inline tchannel/errors.py]: Error frame has trace field

File: tchannel/tornado/connection.py (hunk 2 above)
Comment (junchaowu) [inline tchannel/tornado/connection.py]: this is internal api. User won't use it.

File: tchannel/tornado/dispatch.py [hunk 3]
```diff
@@ -220,20 +222,22 @@ def handle_call(self, request, connection):
...
         except Exception as e:
-            msg = "An unexpected error has occurred from the handler"
-            log.exception(msg)
+            log.exception("Unexpected Error: {0}".format(e.message))
 
-            response.set_exception(TChannelError(e.message))
+            error = UnexpectedError(
+                description='An unexpected error occurred from the handler',
+                id=request.id,
+                tracing=request.tracing,
```
Comment (blampe) [inline tchannel/tornado/dispatch.py]: What's going on with all the tracing stuff? Seems like that's a real bugfix that belongs in its own patch with its own tests.

File: tchannel/tornado/connection.py [hunk 4]
```diff
@@ -471,25 +470,21 @@ def serve(self, handler):
...
-        return self._write(
-            ErrorMessage(
-                code=code,
-                description=description,
-                id=message_id,
-            ),
+
+        error_message = build_raw_error_message(error)
+        write_future = self._write(error_message)
+        tornado.ioloop.IOLoop.current().add_future(
```
Comment (blampe) [inline tchannel/tornado/connection.py]: So there's an important difference between `add_future` like this and `add_done_callback`. `add_future` schedules things such that another coroutine can run between our future finishing and an event hook firing. `add_done_callback` will execute the hook immediately after the future finishes.

Since event hooks tend to be concerned about things like timing, logging, etc., I think we actually *do* want to use `add_done_callback` with our event hooks. @abhinav would you agree with that?

File: tchannel/tornado/connection.py (hunk 2 above)
Comment (blampe) [inline tchannel/tornado/connection.py]: You're right, I thought I was looking at an event hook.

File: tchannel/tornado/connection.py [hunk 5]
```diff
@@ -471,25 +470,21 @@ def serve(self, handler):
...
         """Convenience method for writing Error frames up the wire.
 
-        :param code:
-            Error code
-        :param description:
-            Error description
-        :param message_id:
-            Message in response to which this error is being sent
+        :param error:
+            TChannel Error.
```
Comment (blampe) [inline tchannel/tornado/connection.py]: Reference the actual type with :py:class\`path.to.TheClass\`.

File: tchannel/tornado/connection.py (hunk 4 above)
Comment (abhinav) [inline tchannel/tornado/connection.py]: Sounds reasonable

Comment (junchaowu) [inline tchannel/tornado/connection.py]: If your puts time counting things into their event hook, will it results incorrect value?

File: tchannel/tornado/dispatch.py [hunk 6]
```diff
@@ -149,14 +147,18 @@ def handle_call(self, request, connection):
...
 
         if request.endpoint in self.handlers and requested_as != expected_as:
-            connection.send_error(
-                ErrorCode.bad_request,
-                "Your serialization was '%s' but the server expected '%s'" % (
-                    requested_as,
-                    expected_as,
+            connection.send_error(BadRequestError(
+                description=(
+                    "Your serialization was '%s' but the server expected '%s'"
```
Comment (abhinav) [inline tchannel/tornado/dispatch.py]: I would just do,

> Server expected a "%s" request but got "%s"

So it'll end up like 'Server expected "thrift" request but got "json"'.

File: tchannel/tornado/dispatch.py [hunk 7]
```diff
@@ -220,20 +222,22 @@ def handle_call(self, request, connection):
...
-                e.message,
-                request.id,
-            )
+            e.tracing = request.tracing
+            e.id = request.id
+            connection.send_error(e)
         except Exception as e:
-            msg = "An unexpected error has occurred from the handler"
-            log.exception(msg)
+            log.exception("Unexpected Error: {0}".format(e.message))
```
Comment (abhinav) [inline tchannel/tornado/dispatch.py]: We don't use `str.format` anywhere else in the library. Let's use `%`-based formatting here too.

File: tchannel/tornado/connection.py (hunk 4 above)
Comment (blampe) [inline tchannel/tornado/connection.py]: Right.

`a().add_done_callback(b)` -> `a b c`

`add_future(a(), b) -> a c b` (potentially)

File: tchannel/tornado/dispatch.py (hunk 7 above)
Comment (junchaowu) [inline tchannel/tornado/dispatch.py]: I checked this link: http://stackoverflow.com/questions/5082452/python-string-formatting-vs-format

% maybe not good.

Comment (abhinav) [inline tchannel/tornado/dispatch.py]: We've had this discussion before. I'm not aware of any concrete reasons why one is better than the other. 

The reason I'm asking for the change here is solely for consistency.

**Comment by abhinav** [comment]
lgtm
